config/v1/types_cluster_version.go

```diff
@@ -215,6 +222,58 @@ type UpdateHistory struct {
 // ClusterID is string RFC4122 uuid.
 type ClusterID string
 
+// ClusterVersionCapabilitiesDefault is a constant representing the
+// default policy for optional, core cluster components.
+type ClusterVersionCapabilitiesDefault string
+
+const (
+	// IncludeClusterVersionCapabilities indicates that optional,
+	// core cluster components should be included by default.
+	IncludeClusterVersionCapabilities = "Include"
+
+	// ExcludeClusterVersionCapabilities indicates that optional,
+	// core cluster components should be excluded by default.
+	ExcludeClusterVersionCapabilities = "Exclude"
+)
+
+// ClusterVersionCapabilitiesSpec configures the installation of
+// optional, core cluster components.
+// +k8s:deepcopy-gen=true
+type ClusterVersionCapabilitiesSpec struct {
+	// inclusionDefault declares the default policy for optional,
+	// core cluster components.
+
+	// +optional
+	InclusionDefault ClusterVersionCapabilitiesDefault `json:"inclusionDefault,omitempty"`
+
+	// include declares capabilities that should be included,
+	// regardless of inclusionDefault.
+	// +listType=atomic
+	// +optional
+	Include []string `json:"include,omitempty"`
+
+	// exclude declares capabilities that should be excluded,
+	// regardless of inclusionDefault.
+	// +listType=atomic
+	// +optional
+	Exclude []string `json:"exclude,omitempty"`
+}
+
+// ClusterVersionCapabilitiesStatus describes the state of optional,
+// core cluster components.
+// +k8s:deepcopy-gen=true
+type ClusterVersionCapabilitiesStatus struct {
```

Comment:
should inclusionDefault be in here too?  We can't allow a user to change from "defaultInclude" to "defaultExclude", so we need status to reflect what's actually being applied, right?  (Unless we end up w/ a webhook that enforces the immutability of that field)

Reply:
weren't we also going to have some sort of message field that would indicate if the user tried to do something in spec that we were going to ignore?  (such as adding a new cap to the exclude list)

Reply:
> We can't allow a user to change from "defaultInclude" to "defaultExclude", so we need status to reflect what's actually being applied, right?

I'm still not convinced that we need to forbid default changing, as long as we forbid cap removal.  For example, I don't mind users going from:

```yaml
capabilities:
  inclusionDefault: Include
    exclude:
    - capability1
```

to:

```yaml
capabilities:
  inclusionDefault: Exclude
    include:
    - every
    - cap
    - exceptCapability1
```

as long as the resulting `exclude` set doesn't grow new members.  In fact, I expect CVO validation rejecting unrecognized capabilities (`myTypoedCap1`) to cause some users to want to flow:

1. Expect to like all our stuff, so install 4.y with the default `inclusionDefault: Include`.
2. Hear that 4.(y+1) brings some new cap that they don't like.
3. Attempt to add `badNewCap` to `exclude`, but have the CVO reject them "I have never heard of `badNewCap`.  Typo?".
4. Flip to `inclusionDefault: Exclude` with `include` populated with `status.capabilities.include` contents.
5. Update to 4.(y+1), without having `badNewCap` installed.

None of that _prevents_ us from having `inclusionDefault` in the `status` struct too, just in case we want it.  But I'd rather punt it for now, and circle back if we decide that we do want it later.

> weren't we also going to have some sort of message field that would indicate if the user tried to do something in spec that we were going to ignore? (such as adding a new cap to the exclude list)

We have the existing `status.conditions[]`.  As discussed briefly in [my initial comment][1], I was planning on extending this with a new `type`.  There's existing precedent for the CVO defining custom types locally:

```console
$ grep ClusterStatusConditionType pkg/cvo/status.go
        ClusterStatusFailing = configv1.ClusterStatusConditionType("Failing")
const ClusterVersionInvalid configv1.ClusterStatusConditionType = "Invalid"
```

so I was going to punt again, define a new type locally, and then maybe promote that type into openshift/api later if folks felt the ability to share a central Go constant outweighed the awkwardness about having a ClusterVersion-specific `type` in an enum that is aimed more directly at ClusterOperator.

[1]: https://github.com/openshift/api/pull/1106#issue-1120094064

Reply:
i can live w/ the punt for now, as you say it's nothing we can't add later.

Reply:
this thread is pretty far removed from the current state of the proposed api, so i'm going to resolve it and i'll start a new thread if i still have questions.